datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Plumbing command for running a command on each (sub)dataset"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+
+from argparse import REMAINDER
+from itertools import chain
+from tempfile import mkdtemp
+
+from datalad.cmd import NoCapture, StdOutErrCapture
+from datalad.core.local.run import normalize_command
+
+from datalad.distribution.dataset import (
+    Dataset,
+    EnsureDataset,
+    datasetmethod,
+    require_dataset,
+)
+from datalad.dochelpers import exc_str
+from datalad.interface.base import (
+    Interface,
+    build_doc,
+)
+from datalad.interface.common_opts import (
+    jobs_opt,
+    recursion_flag,
+    recursion_limit,
+)
+from datalad.interface.results import get_status_dict
+from datalad.interface.utils import eval_results
+from datalad.support.constraints import (
+    EnsureBool,
+    EnsureNone,
+)
+from datalad.support.parallel import (
+    ProducerConsumer,
+    ProducerConsumerProgressLog,
+    no_parentds_in_futures,
+    no_subds_in_futures,
+)
+from datalad.support.param import Parameter
+from datalad.utils import (
+    SequenceFormatter,
+    getpwd,
+    shortened_repr,
+)
+
+lgr = logging.getLogger('datalad.local.foreach')
+
+
+@build_doc
+class ForEach(Interface):
+    r"""Run a command on the dataset and/or each of its sub-datasets.
+
+    Each dataset's repo runner is reused to execute the commands, so they
+    are executed in the top directory of a corresponding dataset.
+    In contrast, Python commands (expressions) are evaluated without changing directory.
+    TODO: unify via explicit --chdir? but not sure if worth it
+
+    WARNING: Python expressions are `eval`ed within the scope of this process
+    and provided existing objects.
+
+    """
+    # TODO:     _examples_ = [], # see e.g. run
+
+    _params_ = dict(
+        cmd=Parameter(
+            args=("cmd",),
+            nargs=REMAINDER,
+            metavar='COMMAND',
+            doc="""command for execution. A leading '--' can be used to
+            disambiguate this command from the preceding options to DataLad."""),
+        python=Parameter(
+            args=("--python",),
+            action="store_true",
+            doc="""if given, must be a boolean flag indicating whether
+            the `cmd` is a Python expression to be evaluated, instead of
+            an external command to be executed"""),
```

Comment:
by name change you mean `--python` -> `--eval` or `refds` -> `rootds`, or both?

now I am thinking that may be we would come up with additional "ways " or types of "CMD", so may be worthwhile right away to RF `--python` --> `--cmd-type={external,eval}`?  in Python API it could even be `--cmd-type={auto,external,eval,call}` if `CMD` is a callable function to which we would pass all those variables as kwargs, and `auto` would choose `external` if str, and `call` if callable is provided.

Reply:
got an immediate other use-case, `exec` so it is possible if so desired to provide a bit more complete python source code with imports etc, so I will do the `--cmd-type`

Reply:
Have ATM
```
  --cmd-type {external|exec|eval}
                        type of the command. EXTERNAL: to be run in a child
                        process using dataset's runner; 'exec': Python source
                        code to execute using 'exec(), no value returned;
                        'eval': Python source code to evaluate using 'eval()',
                        return value is placed into 'result' field.
                        Constraints: value must be one of ('external', 'exec',
                        'eval') [Default: 'external']
```
May be later we would like to add some Python API specific mode to just pass a callable which will be given all the placeholders as kwargs.  But I do not think we even support such ability ATM for Python-specific constraint, do we?